engine/src/main/java/org/terasology/world/block/IBlockRegion.java

```diff
@@ -0,0 +1,142 @@
+// Copyright 2020 The Terasology Foundation
+// SPDX-License-Identifier: Apache-2.0
+package org.terasology.world.block;
+
+import org.joml.AABBf;
+import org.joml.LineSegmentf;
+import org.joml.Matrix4fc;
+import org.joml.Planef;
+import org.joml.Rayf;
+import org.joml.Spheref;
+import org.joml.Vector2f;
+import org.joml.Vector3f;
+import org.joml.Vector3fc;
+import org.joml.Vector3i;
+import org.joml.Vector3ic;
+import org.terasology.entitySystem.entity.EntityRef;
+
+import java.util.Iterator;
+import java.util.Optional;
+
+public interface IBlockRegion extends Iterable<Vector3ic> {
+    public static BlockRegion invalid();
+
+    // -- CONSTRUCTORS ----------------------------------------------------------------------------------------------//
+    BlockRegion();
+    public BlockRegion(int minX, int minY, int minZ, int maxX, int maxY, int maxZ);
+    public BlockRegion(Vector3ic min, Vector3ic max);
+    public BlockRegion(int x, int y, int z);
+    public BlockRegion(Vector3ic block);
+    public BlockRegion(BlockRegion source)
+
+    // -- ITERABLE -------------------------------------------------------------------------------------------------//
+    public Iterator<Vector3ic> iterator();
+
+    // -- GETTERS & SETTERS -----------------------------------------------------------------------------------------//
+    public BlockRegion set(int minX, int minY, int minZ, int maxX, int maxY, int maxZ);
+    public BlockRegion set(Vector3ic min, Vector3ic max);
+    public BlockRegion set(BlockRegion source);
+
+    // -- min -------------------------------------------------------------------------------------------------------//
+    public int minX();
+    public BlockRegion minX(int x);
+
+    public int minY();
+    public BlockRegion minY(int y);
+
+    public int minZ();
+    public BlockRegion minZ(int z);
+
+    public Vector3i getMin(Vector3i dest);
+
+    public BlockRegion setMin(int x, int y, int z);
+    public BlockRegion setMin(Vector3ic min);
+    public BlockRegion translateMin(int dx, int dy, int dz);
+    public BlockRegion translateMin(Vector3ic dmin);
+
+    // -- max -------------------------------------------------------------------------------------------------------//
+    public int maxX();
+    public BlockRegion maxX(int x);
+
+    public int maxY();
+    public BlockRegion maxY(int y);
+
+    public int maxZ();
+    public BlockRegion maxZ(int z);
+
+    public Vector3i getMax(Vector3i dest);
+
+    public BlockRegion setMax(int x, int y, int z);
+    public BlockRegion setMax(Vector3ic max);
+    public BlockRegion translateMax(int dx, int dy, int dz);
```

Comment:
Why do we need translation for min and max. Operation is easily achieved with getting the min with with vector and doing an add.  Just rename it to addMin/addMax

Reply:
I also considered `addMax` (and potentially `subMax`) as alternatives here. I'm fine with renaming it :+1: 

The use case I had in mind here was to extend a given region upwards by a few blocks:
```java
region.addMax(0,2,0)
```
as opposed to
```java
region.setMax(region.maxX(), region.maxY() + 2, region.maxZ())
// .. or when instantiating a temporary Vector3i
region.setMax(region.getMax(new Vector3i()).add(0,2,0))
```

I think this is quite verbose and the use case of extending only in a specific direction seems reasonable to me.

Reply:
That makes sense. I don't think moving a corner is really a translation of a blockregion. It's more of a matter of semantics.

Reply:
> It's a translation but add is simpler to understand.

Is it though? What about the ambiguity of add a max and add _to_ a max?

Reply:
> I don't think moving a corner is really a translation of a blockregion.

It's not a translation of a blockregion, but a translation of the max point.

Reply:
> I don't think moving a corner is really a translation of a blockregion
Yep, it's rather only translating one of the corners...

> What about the ambiguity of adding a (max) point
Hm, I'd think this would only be ambiguous if block regions would have a notion of adding things to it (under that name).

Reply:
Sounds strange to me to refer to it as a translation. I guess it makes sense.